File: leetcode/2024_8_august/python/08_spiral_matrix_iii.py
# CODE

from typing import List

class Solution:
    def spiralMatrixIII(self, rows: int, cols: int, rStart: int, cStart: int) -> List[List[int]]:
        dirX = [0, 1, 0, -1]
        dirY = [1, 0, -1, 0]
        result = []

        cnt = 1
        n = dir = 0
        while n < rows * cols:
            
            # to generate spiral matrix, we have to follow 
            # sequence 1, 1, 2, 2, 3, 3, 4, 4, ...
            # thus, outside loop runs for 2 times.
            for _ in range(2):
                for _ in range(cnt):
                    if (rStart >= 0 and rStart < rows and cStart >= 0 and cStart < cols):
                        result.append([rStart, cStart])
                    rStart += dirX[dir]
                    cStart += dirY[dir]
                    
                # generate next direction
                dir = (dir + 1) % 4
            cnt += 1
            n = len(result)
        
        return result

File: leetcode/2024_8_august/c++/08_spiral_matrix_iii.cpp
// TIME COMPEXITY
// O(max(rows,cols) ^ 2)

// SPACE COMPEXITY
// O(1)

// CODE

class Solution {
public:
    vector<vector<int>> spiralMatrixIII(int rows, int cols, int rStart, int cStart) {
        vector<vector<int>> direction = {{0,1}, {1,0}, {0,-1}, {-1,0}}; // E S W N
        vector<vector<int>> res;


        int steps = 0; // increase on East and West
        int dir = 0; // currently east

        res.push_back({rStart, cStart});

        while(res.size() < rows * cols){
            if(dir == 0 || dir == 2){ // East and West
                steps++;
            }
            for(int i = 0; i < steps; i++){// jitne steps h utna chalna h
                rStart += direction[dir][0]; // row increament
                cStart += direction[dir][1]; // col increament

                // if valid
                if(rStart >= 0 && rStart < rows && cStart >= 0 && cStart < cols){
                    res.push_back({rStart, cStart});
                }
            }

            dir = (dir + 1) % 4;
        }

        return res;
    }
};